README.md

```diff
@@ -399,235 +391,177 @@ function runQuery(query) {
+    const port = server.address().port;
+    console.log('Example app listening at http://%s:%s', host, port);
+  });
+}).catch(function(err) {
+  console.error('Error creating connection pool', err);
+});
 ```
 
-All values are automatically sanitized against sql injection.
-
-### Managing connection pools
-
-Most applications will only need a single `ConnectionPool` that can be shared throughout the code. To aid the sharing
-of a single pool this library exposes a set of functions to access a single global connection. eg:
+Then the route uses the passed global connection pool object.
 
 ```js
-// as part of your application's boot process
+// ./routes/route1.js
+const sql = require('mssql');
+
+module.exports = function(globalPool) {
+  const me = {
+    get: function(req, res, next) {
+      const request = new sql.Request(globalPool);
+      request.query('SELECT TOP 10 * FROM table_name', function(err, recordset) {
+        if (err) {
+          console.error(err);
+          res.status(500).send(err.message);
+          return;
+        }
+        res.status(200).json(recordset);
+      });
+    }
+  };
 
-const sql = require('mssql')
-const poolPromise = sql.connect()
+  return me;
+};
+```
 
-// during your applications runtime
+### Advanced Pool Management
 
-poolPromise.then(() => {
-  return sql.query('SELECT 1')
-}).then(result => {
-  console.dir(result)
-})
+For some use-cases you may want to implement your own connection pool management, rather than using the global connection pool. Reasons for doing this include:
 
-// when your application exits
-poolPromise.then(() => {
-  return sql.close()
-})
-```
+* Supporting connections to multiple databases
+* Creation of separate pools for read vs read/write operations
 
-If you require multiple pools per application (perhaps you have many DBs you need to connect to or you want a read-only
-pool), then you will need to manage your pools yourself. The best way to do this is to create a shared library file that
-can hold references to the pools for you. For example:
+The following code is an example of a custom connection pool implementation.
 
 ```js
-const sql = require('mssql')
+// CustomPoolManager.js
+const { ConnectionPool } = require('mssql')
+const POOLS = {}
 
-const pools = {}
-
-// manage a set of pools by name (config will be required to create the pool)
-// a pool will be removed when it is closed
-async function getPool(name, config) {
-  if (!Object.prototype.hasOwnProperty.call(pools, name)) {
-    const pool = new sql.ConnectionPool(config)
-    const close = pool.close.bind(pool)
-    pool.close = (...args) => {
-      delete pools[name]
-      return close(...args)
-    }
-    await pool.connect()
-    pools[name] = pool
+async function createPool(config, name) {
+  if (getPool(name)) {
+    return Promise.reject(new Error('Pool with this name already exists'))
   }
-  return pools[name]
+  const pool = new ConnectionPool(config)
```

Comment:
You've removed a decent amount of the code in this example that actually makes the management of these pools resilient.

With your code, if I call `createPool()` multiple times (in parallel) then you're going to create race conditions and actually create multiple pools that will then be orphaned and never closed:

```js
// this will create 4 different pools, none will throw.
Promise.all([
  createPool(config, 'test'),
  createPool(config, 'test'),
  createPool(config, 'test'),
  createPool(config, 'test'),
]);
```

This also removes the ability to call `pool.close()` on the pool itself and have the reference to it removed from the stored pools. I think it's pretty unintuitive to stop users from being able to use `pool.close()` to remove pools from the store.

Reply:
Looking at this, I think the existing code also suffers from the race conditions when creating pools.

This is probably a more simple yet resilient implementation:

```js
const mssql = require('mssql')
const pools = new Map();

module.exports = {
  /**
   * Get or create a pool. If a pool doesn't exist the config must be provided.
   * If the pool does exist the config is ignored (even if it was different to the one provided
   * when creating the pool)
   *
   * @param {string} name
   * @param {{}} [config]
   * @return {Promise.<mssql.ConnectionPool>}
   */
  get: (name, config) => {
    if (!pools.has(name)) {
      if (!config) {
        throw new Error('Pool does not exist');
      }
      const pool = new mssql.ConnectionPool(config);
      const close = pool.close.bind(pool);
      pool.close = (...args) => {
        pools.delete(name);
        return close(...args);
      }
      pools.set(name, pool.connect());
    }
    return pools.get(name);
  },
  /**
   * Closes all the pools and removes them from the store
   *
   * @return {Promise<mssql.ConnectionPool[]>}
   */
  closeAll: () => Promise.all(Array.from(pools).map(([, connect]) => {
    return connect.then((pool) => pool.close());
  })),
};
```